### ISSUE TITLE
no-unused-imports rule fails to detect certain unused interfaces

### ISSUE THREAD

**Issue by kenlyon** [sweepline/eslint-plugin-unused-imports#66]
I recently noticed that certain unused imports were not being detected and consequently not being removed on save by VS Code.

I found that the offending items were _interfaces_ I was importing from `react`.

(The typings used in my environment are defined in `@types/react@18.0.28`.)
```
import { FocusEvent, useCallback, useEffect, useRef, useState } from "react";
```
The rule recognizes the four imported functions as unused, but not the interface:

![image](https://github.com/sweepline/eslint-plugin-unused-imports/assets/10090224/84e1d4a8-175f-46b4-885b-3dab6d41b72d)

![image](https://github.com/sweepline/eslint-plugin-unused-imports/assets/10090224/1d34b49a-6fd4-4fa0-8d3d-fa7fbfbac2e0)

I did try a similar test using a local `.ts` file exporting a variety of things including an interface and that _did work_. I don't know if there's something unusual about the react typings, but I think the plugin should be able to handle them.

**Comment by kenlyon** [comment]
I think it is.

The same `.eslintrc.js` file that references the `"unused-imports"`` plugin also includes:
* `"@typescript-eslint"` in the `plugins` array.
* `"plugin:@typescript-eslint/recommended"` in the `extends` array.
* A number of rules with the `@typescript-eslint/` prefix.

Additionally, I tried changing an `import type` in the same file where I was seeing the problem to a regular `import` and it triggered an error of type `@typescript-eslint/consistent-type-imports`.

**Comment by sweepline** [comment]
The problem is that my package does not actually define the behavior of when something is unused. It is defined in  `eslint/no-unused-vars` or `@typescript-eslint/no-unused-vars`. So if it is not working then it is probably a configuration issue or their issue. Can you provide me a minimal reproducible example?

**Comment by kenlyon** [comment]
I think this is the minimum you will need.

My project was using `yarn`, so I've done the same here. My global npm install looks like this:
 
 ```
C:\>npm list -g
C:\Program Files\nodejs -> .\
+-- corepack@0.20.0
+-- npm@10.1.0
`-- yarn@1.22.21
```
You will need yarn@1.* installed globally.

Steps to reproduce:
1. Unzip [unused-import-bug.zip](https://github.com/sweepline/eslint-plugin-unused-imports/files/14502009/unused-import-bug.zip) and open it in VS Code
2. In the terminal, run `yarn`.
3. Open `src/test.ts`

After eslint has initialized, you see the `eslint(unused-imports/no-unused-imports)` error on all the imports except `FocusEvent`.

**Comment by sweepline** [comment]
If you change the `.eslintrc.cjs` to the following
```cjs
module.exports = {
    parser: "@typescript-eslint/parser",
    parserOptions: {
        ecmaVersion: "latest",
        project: ["./tsconfig.json"],
        sourceType: "script",
    },
    plugins: ["unused-imports", "@typescript-eslint"],
    rules: {
        "@typescript-eslint/no-unused-vars": "error",
    },
};
```
You can confirm that the same behavior happens in the rule `@typescript-eslint/no-unused-vars` which this package just composes rules from.

 Therefore you should report the issue [upstream](https://github.com/typescript-eslint/typescript-eslint).

**Comment by kenlyon** [comment]
@sweepline Thanks for confirming this. I'll follow up there instead.

**Comment by kenlyon** [comment]
In case anyone's interested, I got an explanation for this problem.

I logged [this issue](https://github.com/typescript-eslint/typescript-eslint/issues/8631) with @typescript-eslint/no-unused-vars and learned the cause of the problem was a mistake in my configuration.

I needed to use `sourceType: "module"` in my `parserOptions` whenever specifying a typescript project. I had set `sourceType: "script"` in my root `.eslintrc.js` because of some plain `.js` script files in my repository, but I should have overridden it with `sourceType: "module"` in my typescript workspaces.
